### PR TITLE
[feat] Added more augmentations containers

### PR DESCRIPTION
**Pull request by shijianjian**
I have currently enabled the mixed ```nn.Module``` processing in our ```Sequential``` implementations.

We shall consider to have a series of ```kornia.nn.Sequential``` and ```kornia.nn.VideoSequential``` for the better naming space management.

### CODE REVIEW

**Comment by edgarriba** [comment]
@shijianjian consider to adapt this https://github.com/PyTorchLightning/lightning-flash/blob/master/flash/core/data/transforms.py#L57

**Comment by shijianjian** [comment]
@edgarriba It is already there if you mark the sequence with ["input", "input"]. Alternatively, mask, bbox and keypoints will be available as well. 

```python
>>> aug_list = AugmentationSequential([
        ...     kornia.augmentation.ColorJitter(0.1, 0.1, 0.1, 0.1, p=1.0),
        ...     kornia.augmentation.RandomAffine(360, p=1.0),
        ...     kornia.enhance.Invert()
        ... ],
        ... input_types=["input", "input"],
        ... return_transform=False,
        ... same_on_batch=False,
        ... )
```

**Comment by shijianjian** [comment]
@edgarriba It is already there if you mark the sequence with ["input", "input"]. Alternatively, mask, bbox and keypoints will be available as well. 

```python
>>> aug_list = AugmentationSequential([
        ...     kornia.augmentation.ColorJitter(0.1, 0.1, 0.1, 0.1, p=1.0),
        ...     kornia.augmentation.RandomAffine(360, p=1.0),
        ...     kornia.enhance.Invert()
        ... ],
        ... input_types=["input", "input"],
        ... return_transform=False,
        ... same_on_batch=False,
        ... )
```

File: kornia/augmentation/container/__init__.py [hunk 1]
```diff
@@ -1 +1,3 @@
 from .video import VideoSequential
+from .augmentation_seqential import AugmentationSequential
```
Comment (edgarriba) [inline kornia/augmentation/container/__init__.py:2]:
```suggestion
from .augmentation_sequential import AugmentationSequential
```

File: kornia/augmentation/container/augmentation_seqential.py [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+
+from kornia.geometry import transform_points, transform_boxes
+from kornia.augmentation.base import (
+    _AugmentationBase,
+    IntensityAugmentationBase2D,
+    GeometricAugmentationBase2D,
+)
+from .sequential import Sequential
+
+SUPPORTED_INPUT_TYPES = ["input", "mask", "bbox", "bbox_xyxy", "bbox_xywh", "keypoints"]
```
Comment (edgarriba) [inline kornia/augmentation/container/augmentation_seqential.py:14]: I'd suggest to have here enum types to avoid typo errors

**Review by edgarriba — CHANGES_REQUESTED** [review]

File: kornia/constants.py [hunk 3]
```diff
@@ -21,20 +21,41 @@ def get(cls, value: Union[str, int, T]) -> T:  # type: ignore
...
 
-class SamplePadding(ConstantBase, Enum):
+class SamplePadding(ConstantBase, Enum, metaclass=EnumMetaFlags):
     ZEROS = 0
     BORDER = 1
     REFLECTION = 2
+
+
+class InputType(ConstantBase, Enum, metaclass=EnumMetaFlags):
+    INPUT = 0
```
Comment (edgarriba) [inline kornia/constants.py:56]: not huge fan of this naming, a bit redundant to have `InputType` and then `INPUT` referring to images, feature tensors, etc. I would suggest something like `DataType`

File: kornia/augmentation/__init__.py [hunk 4]
```diff
@@ -40,7 +40,7 @@
     RandomPerspective3D,
 )
 from .mix_augmentation import RandomMixUp, RandomCutMix
-from .container import VideoSequential
+from .container import VideoSequential, AugmentationSequential, Sequential
```
Comment (edgarriba) [inline kornia/augmentation/__init__.py:43]: I think we should start considering to add full path imports

File: kornia/augmentation/__init__.py [hunk 5]
```diff
@@ -81,6 +81,11 @@
     "RandomCrop3D",
     "CenterCrop3D",
     "RandomEqualize3D",
-    "VideoSequential",
     "RandomThinPlateSpline",
+    "RandomChannelShuffle",
```
Comment (edgarriba) [inline kornia/augmentation/__init__.py:85]: Alphabetical order better

File: kornia/augmentation/container/augmentation_sequential.py [hunk 6]
```diff
@@ -0,0 +1,268 @@
+from typing import Tuple, Union, Optional, List, Dict, cast
+
+import torch
+import torch.nn as nn
+
+from kornia.geometry import transform_points, transform_boxes
+from kornia.augmentation.base import _AugmentationBase, IntensityAugmentationBase2D, GeometricAugmentationBase2D
+from kornia.constants import DataCategory
+from .sequential import Sequential
```
Comment (edgarriba) [inline kornia/augmentation/container/augmentation_sequential.py:9]: I would avoid confusion naming with `torch.nn.Sequential`

File: kornia/augmentation/container/augmentation_sequential.py [hunk 7]
```diff
@@ -0,0 +1,268 @@
...
+
+
+class AugmentationSequential(Sequential):
+    r"""AugmentationSequential for handling multiple input types like inputs, masks, keypoints at once.
+
+    Args:
+        *args (_AugmentationBase): a list of augmentation module.
```
Comment (edgarriba) [inline kornia/augmentation/container/augmentation_sequential.py:16]:
```suggestion
        *args (_AugmentationBase): a list of augmentation modules.
```

File: kornia/augmentation/container/augmentation_sequential.py [hunk 8]
```diff
@@ -0,0 +1,268 @@
...
+
+        assert all(
+            [in_type in DataCategory for in_type in self.input_types]
+        ), f"`input_types` must be in {DataCategory}. Got {input_types}."
+
+        if self.input_types[0] != DataCategory.INPUT:
+            raise NotImplementedError(f"The first input must be {DataCategory.INPUT}.")
+
+    def apply_to_mask(
+        self, input: torch.Tensor, item: nn.Module, param: Optional[Dict[str, torch.Tensor]] = None
```
Comment (edgarriba) [inline kornia/augmentation/container/augmentation_sequential.py:77]:
```suggestion
        self, input: torch.Tensor, transform: nn.Module, params: Optional[Dict[str, torch.Tensor]] = None
```

File: kornia/augmentation/container/augmentation_sequential.py [hunk 9]
```diff
@@ -0,0 +1,268 @@
...
+        return input
+
+    def apply_to_keypoints(
+        self, input: torch.Tensor, item: nn.Module, param: Optional[Dict[str, torch.Tensor]] = None
+    ) -> torch.Tensor:
+        if isinstance(item, GeometricAugmentationBase2D) and param is None:
+            raise ValueError(f"Transformation matrix for {item} has not been computed.")
+        elif isinstance(item, GeometricAugmentationBase2D) and param is not None:
+            input = transform_points(
+                torch.as_tensor(item.get_transformation_matrix(input, param), device=input.device, dtype=input.dtype),
```
Comment (edgarriba) [inline kornia/augmentation/container/augmentation_sequential.py:109]: get_transform should come as tensor already

File: kornia/augmentation/container/augmentation_sequential.py [hunk 10]
```diff
@@ -0,0 +1,268 @@
...
+                input,
+            )
+        else:
+            pass  # No need to update anything
+        return input
+
+    def apply_by_input_type(
```
Comment (edgarriba) [inline kornia/augmentation/container/augmentation_sequential.py:116]:
```suggestion
    def apply_by_key(
```

File: kornia/augmentation/container/augmentation_sequential.py [hunk 11]
```diff
@@ -0,0 +1,268 @@
...
+                else:
+                    raise NotImplementedError(f"input_type {itype} is not implemented for {item}.")
+            outputs.append(input)
+
+        if len(outputs) == 1:
+            return outputs[0]
+
+        return outputs
+
+    def forward(  # type: ignore
```
Comment (edgarriba) [inline kornia/augmentation/container/augmentation_sequential.py:232]: Why type ignore?

File: kornia/augmentation/container/sequential.py [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+            else:
+                input = item(input)
+        return input
+
+    def forward(
+        self, input: torch.Tensor, params: Optional[Dict[str, Dict[str, torch.Tensor]]] = None
+    ) -> Union[torch.Tensor, Tuple[torch.Tensor, torch.Tensor]]:
+        if params is None:
+            params = {}
+        self._params = {}
```
Comment (edgarriba) [inline kornia/augmentation/container/sequential.py:96]: Where do you use this?

File: kornia/augmentation/container/sequential.py [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+        for item in self.children():
+            func_name = item.__class__.__name__
+            param = params[func_name] if func_name in params else None
+            input = self.apply_to_input(input, item, param)  # type: ignore
```
Comment (edgarriba) [inline kornia/augmentation/container/sequential.py:100]: Why type ignore?

File: kornia/constants.py [hunk 14]
```diff
@@ -21,20 +21,42 @@ def get(cls, value: Union[str, int, T]) -> T:  # type: ignore
         raise TypeError()
 
 
-class Resample(ConstantBase, Enum):
+class EnumMetaFlags(EnumMeta):
```
Comment (edgarriba) [inline kornia/constants.py:24]: Why this class needed?

**Review by edgarriba — CHANGES_REQUESTED** [review]

File: kornia/constants.py (hunk 14 above)
Comment (shijianjian) [reply kornia/constants.py:24]: Then we can use things like ```"mask" in DataCategory```.

File: kornia/augmentation/container/sequential.py (hunk 13 above)
Comment (shijianjian) [reply kornia/augmentation/container/sequential.py:100]: Only tensor is allowed for ```forward```, while ```self.apply_to_input ``` also solves ```return_transform=True```. mypy does not aware the potential of input type overriding.

File: kornia/augmentation/container/sequential.py (hunk 12 above)
Comment (shijianjian) [reply kornia/augmentation/container/sequential.py:96]: To forward the params of the whole sequential.

File: kornia/augmentation/container/augmentation_sequential.py (hunk 11 above)
Comment (shijianjian) [reply kornia/augmentation/container/augmentation_sequential.py:232]: The function signature differs from the forward in ```nn.Module```. it complains.

File: kornia/augmentation/__init__.py (hunk 4 above)
Comment (shijianjian) [reply kornia/augmentation/__init__.py:43]: There was a fullpath import error with python==3.6. Full path import will result in some errors with the module initialisation orders.

File: kornia/augmentation/container/augmentation_sequential.py (hunk 6 above)
Comment (shijianjian) [reply kornia/augmentation/container/augmentation_sequential.py:9]: what do you have in mind?

File: kornia/augmentation/container/augmentation_sequential.py [hunk 15]
```diff
@@ -85,17 +85,16 @@ def apply_to_mask(
         return input
 
     def apply_to_bbox(
-        self, input: torch.Tensor, module: nn.Module, param: Optional[Dict[str, torch.Tensor]] = None,
-        mode: str = "xyxy"
+        self,
+        input: torch.Tensor,
+        module: nn.Module,
+        param: Optional[Dict[str, torch.Tensor]] = None,
```
Comment (edgarriba) [inline kornia/augmentation/container/augmentation_sequential.py:91]: In the previous api we were naming this in plural, right ? `params`

**Review by edgarriba — COMMENTED** [review]
Let’s brainstorm a bit more about the container namigs and make sure the coverage warnings do not appear. The rest looks good, I like this api

File: kornia/augmentation/container/augmentation_sequential.py (hunk 15 above)
Comment (shijianjian) [reply kornia/augmentation/container/augmentation_sequential.py:91]: It has ~~~params~~~ in the forward method. So that for those functions that applies one nn.Module, I used singular to differentiate.

**Review by edgarriba — APPROVED** [review]

File: kornia/augmentation/__init__.py [hunk 16]
```diff
@@ -52,33 +52,40 @@
...
+    "RandomPosterize",
     "RandomEqualize",
     "RandomMotionBlur",
+    "RandomInvert",
+    "RandomThinPlateSpline",
     "RandomMixUp",
     "RandomCutMix",
     "AugmentationBase3D",
+    "CenterCrop3D",
+    "Normalize3D",
```
Comment (edgarriba) [inline kornia/augmentation/__init__.py:77]:
```suggestion
```

File: kornia/augmentation/__init__.py [hunk 17]
```diff
@@ -52,33 +52,40 @@
...
+    "Denormalize3D",
```
Comment (edgarriba) [inline kornia/augmentation/__init__.py:78]:
```suggestion
```

**Review by edgarriba — APPROVED** [review]

File: docs/source/augmentation.container.rst [hunk 18]
```diff
@@ -3,16 +3,39 @@ Augmentation Containers
...
+.. autoclass:: AugmentationSequential
 
    .. automethod:: forward
 
+   .. automethod:: inverse
+
+
+
+ImageSequential
+-----------------------
```
Comment (edgarriba) [inline docs/source/augmentation.container.rst:26]:
```suggestion
---------------
```
